src/lib/sync_handler/sync_handler.ml

```diff
@@ -56,24 +64,50 @@ module Make (Inputs : Inputs_intf) :
   let get_staged_ledger_aux_and_pending_coinbases_at_hash ~frontier state_hash
       =
     let open Option.Let_syntax in
-    Option.merge
-      (let%map breadcrumb = Transition_frontier.find frontier state_hash in
-       let staged_ledger =
-         Transition_frontier.Breadcrumb.staged_ledger breadcrumb
-       in
-       let scan_state = Staged_ledger.scan_state staged_ledger in
-       let merkle_root =
-         Staged_ledger.hash staged_ledger |> Staged_ledger_hash.ledger_hash
-       in
-       let pending_coinbase =
-         Staged_ledger.pending_coinbase_collection staged_ledger
-       in
-       (scan_state, merkle_root, pending_coinbase))
-      (let%map root = find_in_root_history frontier state_hash in
-       ( root.scan_state
-       , root.staged_ledger_target_ledger_hash
-       , root.pending_coinbase ))
-      ~f:Fn.const
+    let protocol_states scan_state =
+      Staged_ledger.Scan_state.required_state_hashes scan_state
+      |> State_hash.Set.to_list
+      |> List.fold_until ~init:(Some [])
+           ~f:(fun acc hash ->
+             match
+               acc
+               >>= fun ps ->
+               Transition_frontier.find_protocol_state frontier hash
+               >>= fun p -> Some (p :: ps)
```

Comment:
Nit: This is pretty awkward to read, especially wrapped in a match. Would be better to use `Option.map2`:
```suggestion
Option.map2 ~f:List.cons (Transition_frontier.find_protocol_state frontier hash) acc
```